diff --git a/include/private/gcconfig.h b/include/private/gcconfig.h
index 8e09055..79be594 100644
--- a/include/private/gcconfig.h
+++ b/include/private/gcconfig.h
@@ -725,6 +725,13 @@
 #   define mach_type_known
 # endif
 
+# if defined(__wasi__)
+  /* The WebAssembly System Interface (WASI). */
+#   define WEBASSEMBLY
+#   define WASI
+#   define mach_type_known
+# endif
+
 /* Feel free to add more clauses here */
 
 /* Or manually define the machine type here.  A machine type is         */
@@ -1458,6 +1465,34 @@
 #   endif
 # endif
 
+# ifdef WEBASSEMBLY
+#   define MACH_TYPE "WEBASSEMBLY"
+#   if defined(__wasm64__) && !defined(CPPCHECK)
+#     error 64-bit WebAssembly is not yet supported
+#   endif
+#   define CPP_WORDSZ 32
+#   define ALIGNMENT 4
+#   undef USE_MMAP
+#   undef USE_MUNMAP
+#   ifdef WASI
+#     define OS_TYPE "WASI"
+      extern char __global_base, __data_end, __heap_base;
+#     define DATASTART ((ptr_t)(&__global_base))
+#     define DATAEND ((ptr_t)(&__data_end))
+#     define STACKBOTTOM ((ptr_t)(&__heap_base))
+#     ifndef GC_NO_SIGSETJMP
+#       define GC_NO_SIGSETJMP 1
+#     endif
+#     ifndef NO_CLOCK
+#       define NO_CLOCK 1
+#     endif
+#     define GETPAGESIZE() 65536
+#     if defined(GC_THREADS) && !defined(CPPCHECK)
+#       error No threads support yet
+#     endif
+#   endif
+# endif
+
 # ifdef I386
 #   define MACH_TYPE "I386"
 #   if (defined(__LP64__) || defined(_WIN64)) && !defined(CPPCHECK)
diff --git a/include/private/gc_priv.h b/include/private/gc_priv.h
index 8bd9e5c..f5a8b49 100644
--- a/include/private/gc_priv.h
+++ b/include/private/gc_priv.h
@@ -824,7 +824,9 @@
 # endif
 #endif /* DARWIN */
 
-#include <setjmp.h>
+#ifndef WASI
+# include <setjmp.h>
+#endif
 
 #if defined(CAN_HANDLE_FORK) && defined(GC_PTHREADS)
 #  include <pthread.h> /* for pthread_t */
diff --git a/alloc.c b/alloc.c
index 71119b9..58d3b7f 100644
--- a/alloc.c
+++ b/alloc.c
@@ -19,7 +19,7 @@
 #include "private/gc_priv.h"
 
 #include <stdio.h>
-#if !defined(MACOS) && !defined(MSWINCE)
+#if !defined(MACOS) && !defined(MSWINCE) && !defined(WASI)
 # include <signal.h>
 # if !defined(GC_NO_TYPES) && !defined(SN_TARGET_PSP2) \
      && !defined(__CC_ARM)
diff --git a/misc.c b/misc.c
index 369f638..8db66ad 100644
--- a/misc.c
+++ b/misc.c
@@ -20,7 +20,7 @@
 #include <limits.h>
 #include <stdarg.h>
 
-#ifndef MSWINCE
+#if !defined(MSWINCE) && !defined(WASI)
 # include <signal.h>
 #endif
 
diff --git a/os_dep.c b/os_dep.c
index d8cb91f..bfcad09 100644
--- a/os_dep.c
+++ b/os_dep.c
@@ -27,7 +27,7 @@
 #endif
 
 #include <stdio.h>
-#if defined(MSWINCE) || defined(SN_TARGET_PS3)
+#if defined(MSWINCE) || defined(SN_TARGET_PS3) || defined(WASI)
 # define SIGSEGV 0 /* value is irrelevant */
 #else
 # include <signal.h>
@@ -843,7 +843,7 @@
   GC_INNER void GC_setpagesize(void)
   {
 #   if defined(MPROTECT_VDB) || defined(PROC_VDB) || defined(SOFT_VDB) \
-       || defined(USE_MMAP)
+       || defined(USE_MMAP) || defined(WASI)
       GC_page_size = (size_t)GETPAGESIZE();
 #     if !defined(CPPCHECK)
         if (0 == GC_page_size)
