diff --git a/errno.c b/errno.c
--- a/errno.c
+++ b/errno.c
@@ -34,7 +34,9 @@
 
 #include <stdlib.h>
 
+#if !defined (__wasi__)
 #include <signal.h>
+#endif
 
 #include "gmp-impl.h"
 
@@ -48,7 +50,9 @@
 __gmp_exception (int error_bit)
 {
   gmp_errno |= error_bit;
-#ifdef SIGFPE
+#if defined (__wasi__)
+  abort ();
+#elif defined (SIGFPE)
   raise (SIGFPE);
 #else
   __gmp_junk = 10 / __gmp_0;
diff --git a/invalid.c b/invalid.c
--- a/invalid.c
+++ b/invalid.c
@@ -34,9 +34,12 @@
 
 #include "config.h"
 
-#include <signal.h>
 #include <stdlib.h>
 
+#if !defined (__wasi__)
+#include <signal.h>
+#endif
+
 #if HAVE_UNISTD_H
 #include <unistd.h>  /* for getpid */
 #endif
@@ -46,7 +49,7 @@
 
 /* Incidentally, kill is not available on mingw, but that's ok, it has raise
    and we'll be using that.  */
-#if ! HAVE_RAISE
+#if ! HAVE_RAISE && !defined (__wasi__)
 #define raise(sig)   kill (getpid(), sig)
 #endif
 
@@ -77,6 +80,8 @@
 void
 __gmp_invalid_operation (void)
 {
+#if !defined (__wasi__)
   raise (SIGFPE);
+#endif
   abort ();
 }
