diff --git a/glib/meson.build b/glib/meson.build
index 209bcbf..e49f81b 100644
--- a/glib/meson.build
+++ b/glib/meson.build
@@ -514,7 +514,7 @@ if host_system == 'windows'
       include_directories : configinc,
       dependencies : [libglib_dep])
   endif
-else
+elif host_system != 'wasi'
   gtester = executable('gtester', 'gtester.c',
     install : true,
     install_tag : 'bin-devel',
diff --git a/meson.build b/meson.build
index dab1d61..e61538d 100644
--- a/meson.build
+++ b/meson.build
@@ -2292,7 +2292,11 @@ gvdb_dep = dependency('gvdb')
 libm = cc.find_library('m', required : false)
 libffi_dep = dependency('libffi', version : '>= 3.0.0')
 
-libz_dep = dependency('zlib')
+if host_system == 'wasi'
+  libz_dep = declare_dependency()
+else
+  libz_dep = dependency('zlib')
+endif
 
 libatomic_test_code = '''
   int main (int argc, char ** argv) {
@@ -2720,16 +2724,11 @@ pkg = import('pkgconfig')
 windows = import('windows')
 gnome = import('gnome')
 
-subdir('tools')
+gen_visibility_macros = find_program('tools/gen-visibility-macros.py')
 subdir('glib')
 subdir('gobject')
-subdir('gthread')
-subdir('gmodule')
-subdir('gio')
-subdir('girepository')
-subdir('fuzzing')
-if get_option('tests')
-  subdir('tests')
+if host_system == 'wasi'
+  subdir('gio')
 endif
 
 # xgettext is optional (on Windows for instance)
@@ -2770,7 +2769,6 @@ endif
 rst2html5 = find_program('rst2html5', 'rst2html5.py', required: get_option('documentation'))
 
 gnome = import('gnome')
-subdir('docs/reference')
 
 summary({
   'host cpu' : host_machine.cpu_family(),
